### PR TITLE
WIP: Switch to framework and Mac new app

### PR DESCRIPTION
**Pull request by slarew**
Test the Darwin framework on the buildbot CI by packing it in the DMG.

Putting the framework and app separately in the DMG is not a good distribution solution.  This PR is just to test the framework on the buildbots without needing to change anything on the buildbots. Later, the CI should be updated per #31755.

### CODE REVIEW

**Comment by slarew** [comment]
Progress update: @staticfloat

The mac app (#31754) requires Xcode to build.  Also, the buildbot for macOS appears to be out-of-date (see below).  I see it's running 10.11 El Capitan which is now unsupported.  Is there anything (besides time and effort :) blocking an upgrade of the buildbot to Mojave and Xcode 10?  Note that to support the [notarization requirements](https://developer.apple.com/news/?id=06032019i) in macOS Catalina, Xcode 10 is [required](https://developer.apple.com/documentation/security/notarizing_your_app_before_distribution).

```
xcrun --version
xcrun version 31.
xcrun --show-sdk-path

xcrun --show-sdk-version
10.11
xcrun clang --version
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
sw_vers
ProductName:	Mac OS X
ProductVersion:	10.11.6
BuildVersion:	15G22010
xcode-select -p
/Library/Developer/CommandLineTools
```

**Comment by staticfloat** [comment]
We build on 10.11 simply for backwards compatibility.  It looks like I would need up upgrade the buildbots to High Sierra or later to run XCode 10.  I don't _think_ there will be any problems upgrading to 10.13, so I'll do that on the buildbots when I have some free time.

**Comment by staticfloat** [comment]
The buildbots have now been updated, and they all have Xcode installed; go ahead and try again.  :)

**Comment by DilumAluthge** [comment]
I wanted to post my concerns about this change. Cross-posted from my comment on https://github.com/JuliaLang/julia/issues/31754#issuecomment-517291196

> I'm a little confused by this proposal, and I apologize if I have misunderstood it.
> 
> Would this proposal eliminate the option of downloading a .DMG file that contains a `Julia-1.x.app` file that I can then copy to any location? If so, then I am strongly opposed to it.
> 
> Advantages of DMG with a self contained `Julia-1.x.app` :
> - I can copy `Julia-1.x.app` to any location that I want
> - Copying the `Julia-1.x.app` file to my desired location does not make modifications to other parts of my system.
> - The `Julia-1.x.app` is self contained, and it does not interact with the rest of my system (apart from `~/.julia` of course).
> 
> Disadvantages of macOS .pkg installers:
> - Installers can run arbitrary scripts before, during, and after the installation process.
>     - All I want is a binary build of Julia. Why do I need to let an installer do anything else to my system?
> - Installers will write files to a prespecified set of locations. For example, if the installer is run without admin privileges, applications are installed to `~/Applications` and frameworks are installed to `~/Library/Frameworks`.
>     - Shouldn't I be able to decide exactly where I keep my Julia binaries?
> - In this proposal, the installer would install a macOS Framework.
>     - I've had to deal with Frameworks in the past in the setting of Python. It was not a good experience. I often found that if I deleted, modified, or moved a Framework, then multiple pieces of Python-dependent software on my system would break. Additionally, I often found myself in DLL-hell (or whatever the macOS equivalent is), in which a piece of Python-dependent software had linked itself against one version of a Python Framework, and then when I replaced the Python Framework with a different version, everything broke.
> 
> I definitely understand and appreciate that many people would like the option of having a macOS .pkg installer that installs a macOS Framework. It seems like something that would be beneficial to other people, even if I would personally never use it. Thus, I strongly support the idea of distributing both:
> 1. The macOS .pkg installer that you describe in this post, which will install a Julia.app and a Julia.framework
> 2. A .DMG disk image with a single self-contained `Julia-1.x.app` (the current distribution method)
> 
> So I would strongly urge that we continue to support the current distribution method for macOS users, even if we do add the method proposed here.

**Comment by staticfloat** [comment]
@slarew I have rebased this on latest `master` in an effort to see if the CI issues persist, and I've also taken the liberty of turning the framework off by default.  I understand that this is necessary for some workflows, but I'm not sure it is the default package we want to ship Julia as.  Dilum's concerns above are still valid, and although I believe that frameworks will become necessary if we ever want to ship Julia on iOS, I admit I am not knowledgable enough about macOS development to know what the right way to do that is without diverging too much from the packaging approach taken on other platforms like Linux.

In any case, I appreciate your hard work (and incredible patience) as we've let this PR languish here; as usual, I am happy to have your work in `contrib/mac/framework` for anyone else to make use of, but for now I don't think we're ready to use this as the default distribution format.

**Comment by vtjnash** [comment]
I don't see how this is related, but something seems to be confused:
```diff
diff --git a/Makefile b/Makefile
index f3100ea74d..1c75f297f5 100644
--- a/Makefile
+++ b/Makefile
@@ -410,7 +410,7 @@ endif
                DEBUG_TARGET=$(DESTDIR)$(prefix)/$(framework_dylib)_debug; \
        fi; \
        $(call stringreplace,$${RELEASE_TARGET},sys.$(SHLIB_EXT)$$,$(private_libdir_rel)/sys.$(SHLIB_EXT)); \
-       if [ $(BUNDLE_DEBUG_LIBS) = 1 ]; then \
+       if [ $(BUNDLE_DEBUG_LIBS:-0) = 1 ]; then \
                $(call stringreplace,$${DEBUG_TARGET},sys-debug.$(SHLIB_EXT)$$,$(private_libdir_rel)/sys-debug.$(SHLIB_EXT)); \
        fi;
 endif
```

File: Makefile [hunk 1]
```diff
@@ -488,8 +488,7 @@ exe:
 	chmod a+x "$(BUILDROOT)/$(JULIA_BINARYDIST_FILENAME).exe"
 
 app:
-	$(MAKE) -C contrib/mac/app
-	@mv contrib/mac/app/$(JULIA_BINARYDIST_FILENAME).dmg $(BUILDROOT)
+	$(MAKE) -C contrib/mac/frameworkapp APPLE_DISTRIBUTE_DEVID=1 dmg
```
Comment (vtjnash) [inline Makefile:491]:
```suggestion
ifneq ($(DARWIN_FRAMEWORK),1)
	$(MAKE) -C contrib/mac/app
 	@mv contrib/mac/app/$(JULIA_BINARYDIST_FILENAME).dmg $(BUILDROOT)
 else
	$(MAKE) -C contrib/mac/frameworkapp APPLE_DISTRIBUTE_DEVID=1 dmg
endif
```

File: contrib/mac/frameworkapp/Makefile [hunk 2]
```diff
@@ -17,15 +17,34 @@ XCEXPORT_METHOD ?= development
 # Identity for the product archive (the .pkg installer).
 DARWIN_CODESIGN_PRODUCT_ARCHIVE_KEYCHAIN_IDENTITY ?= -
 else
-DARWIN_CODESIGN_KEYCHAIN_IDENTITY ?= Developer ID Application
-APPLE_DEVELOPMENT_TEAM ?= unidentified
+DARWIN_CODESIGN_KEYCHAIN_IDENTITY ?= AFB379C0B4CBD9DB9A762797FC2AB5460A2B0DBE
+APPLE_DEVELOPMENT_TEAM ?= A427R7F42H
```
Comment (vtjnash) [inline contrib/mac/frameworkapp/Makefile:21]:
```suggestion
DARWIN_CODESIGN_KEYCHAIN_IDENTITY ?= Developer ID Application
APPLE_DEVELOPMENT_TEAM ?= unidentified
```
Or, what are these other numbers?

**Comment by ViralBShah** [comment]
While I really was keen to adopt this at some point, it greatly increases the complexity for maintenance and we are unlikely to adopt this. Closing as a result.

**Pull request closed without merging**